### PR TITLE
Move memory related classes from foundation/utility to memory

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/foundation/array/arrayallocator.h [hunk 1]
```diff
@@ -30,7 +30,7 @@
 
 // appleseed.foundation headers.
 #include "foundation/math/scalar.h"
-#include "foundation/utility/memory.h"
+#include "foundation/memory/memory.h"
```
Comment (dictoon) [inline src/appleseed/foundation/array/arrayallocator.h]: Maybe we could rename `memory.h`?

File: src/appleseed/foundation/platform/python.h [hunk 2]
```diff
@@ -39,15 +39,15 @@
 #endif
 
 // Boost headers.
-#include "boost/python/detail/wrap_python.hpp"  // has to be first, to avoid redefinition warnings
 #include "boost/python.hpp"
+#include "boost/python/detail/wrap_python.hpp"  // has to be first, to avoid redefinition warnings
```
Comment (dictoon) [inline src/appleseed/foundation/platform/python.h]: This change should be reverted I think.

My guess is that you wrote a Python script to reorder the `#include` statements? ;) If so, could you share it?

File: src/appleseed/foundation/utility/minizip/ioapi.h [hunk 3]
```diff
@@ -41,9 +41,9 @@
 
 #endif
 
+#include "zlib.h"
```
Comment (dictoon) [inline src/appleseed/foundation/utility/minizip/ioapi.h]: Revert changes to third party files.

**Comment by dictoon** [comment]
Thanks, looks great! You can merge once you've addressed the minor issues I brought up.

File: src/appleseed/foundation/platform/python.h (hunk 2 above)
Comment (est77) [reply src/appleseed/foundation/platform/python.h]: I used the sort includes script in the repo. But it does not work very well in some cases.
For example it messes with _begin_xxx and _end_xxx headers.

File: src/appleseed/foundation/array/arrayallocator.h (hunk 1 above)
Comment (est77) [reply src/appleseed/foundation/array/arrayallocator.h]: I prefer to leave it like this. having memory.h and memory/ with some other headers looks a bit arbitrary to me.

Comment (dictoon) [reply src/appleseed/foundation/array/arrayallocator.h]: I'm not sure to understand your point :) Mine was: now that we have a `memory` subfolder, `memory.h` isn't very descriptive anymore and maybe we could rename it. Or not. Just wondering.

**Pull request merged**